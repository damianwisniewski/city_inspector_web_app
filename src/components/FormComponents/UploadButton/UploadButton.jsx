import React, { Component } from 'react'
import PropTypes from 'prop-types'

import './UploadButton.scss'
import Loader from '../../CommonComponents/Loader/Loader'
import Status from '../../CommonComponents/Status/Status'

class UploadButton extends Component {
	state = {
		onDropArea: false,
		loading: false,
		loadingFromDropEvent: false,
		loaded: false,
		files: [],
		error: '',
	}

	initImageLoaders = files => {
		const { acceptsFile } = this.props

		const imageLoaders = []

		if (!files.length) {
			// IE11 for catalog instead of file will have file equal empty array
			imageLoaders.push(Promise.reject('Wybrano niepoprawny plik!'))
		} else {
			for (let i = 0; i < files.length; i++) {
				const file = files[i]

				if (file.type.startsWith(acceptsFile)) {
					imageLoaders.push(
						new Promise((resolve, reject) => {
							const reader = new FileReader()

							reader.addEventListener('load', event => {
								resolve(<img src={event.target.result} alt={file.name} />)
							})

							reader.addEventListener('error', () => {
								reject('Przepraszamy wystąpił błąd!')
							})

							reader.readAsDataURL(file)
						}),
					)
				} else {
					imageLoaders.push(Promise.reject('Wybrano niepoprawny plik!'))
				}
			}
		}

		return imageLoaders
	}

	getFiles = files => {
		const imageLoaders = this.initImageLoaders(files)

		Promise.all(imageLoaders)
			.then(images => {
				this.setState({
					loading: false,
					loadingFromDropEvent: false,
					loaded: true,
					error: '',
					files: images,
				})

				if (this.props.onAddImages) {
					this.props.onAddImages(images)
				}
			})
			.catch(errorMessage => {
				this.setState({
					error: errorMessage,
					loading: false,
					loaded: false,
					loadingFromDropEvent: false,
				})
			})
	}

	/**
	 * To prevent IE11 reloading page with image as content
	 */
	handleDragOver = e => {
		e.preventDefault()
		e.stopPropagation()
	}

	outOfDropArea = () => {
		this.setState({ onDropArea: false })
	}

	inDropArea = () => {
		this.setState({ onDropArea: true })
	}

	handleDrop = e => {
		const files = e.dataTransfer.files

		this.setState({
			onDropArea: false,
			loading: true,
			loadingFromDropEvent: true,
			loaded: false,
		})

		this.getFiles(files)
	}

	handleOnChange = e => {
		if (this.state.loadingFromDropEvent) return
		const files = e.target.files

		this.setState({
			onDropArea: false,
			loading: true,
			loadingFromDropEvent: true,
			loaded: false,
		})

		this.getFiles(files)
	}

	render() {
		// eslint-disable-next-line no-unused-vars
		const { id, label, onAddImages, acceptsFile, ...rest } = this.props
		const { onDropArea, loaded, loading, files, error } = this.state

		return (
			<div className='input-wrapper input-wrapper--upload'>
				<input
					{...rest}
					ref={input => (this.input = input)}
					accept='image/*'
					onDrop={this.handleDrop}
					onDragEnter={this.inDropArea}
					onDragLeave={this.outOfDropArea}
					onDragOver={this.handleDragOver}
					onChange={this.handleOnChange}
					type='file'
					multiple
					className='input-wrapper__field'
					id={id}
				/>
				<label
					className={`input-wrapper__upload-label ${
						onDropArea ? 'input-wrapper__upload-label--active' : ''
					}`}
					htmlFor={id}
				>
					{label}
					<p className='input-wrapper__upload-instruction'>Kliknij w pole lub przeciągnij plik</p>
					<div className='input-wrapper__upload-status'>
						{error && <Status type='error' message={this.state.error} />}
						{loading && <Loader id={`${id}-loader`} size='small' />}
						{loaded && (
							<Status
								type='correct'
								message={`Załadowano ${files.length} ${
									files.length === 1 ? 'plik' : files.length <= 4 ? 'pliki' : 'plików'
								}`}
							/>
						)}
					</div>
				</label>
			</div>
		)
	}
}

UploadButton.propTypes = {
	label: '',
}

UploadButton.propTypes = {
	id: PropTypes.string.isRequired,
	label: PropTypes.string,
	onAddImages: PropTypes.func.isRequired,
	acceptsFile: PropTypes.string.isRequired,
}

export default UploadButton
