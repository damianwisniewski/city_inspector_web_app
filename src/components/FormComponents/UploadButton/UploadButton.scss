@import '../../../assets/styles/index';

.input-wrapper--upload {
	.input-wrapper__upload-label {
		position: absolute;
		top: 35px;
		left: 35px;
		z-index: 10;
		width: calc(100% - 70px);
		height: 100px;
		border: 1px dashed $dark-gray;
		background-color: $white;
		border-radius: 3px;
		font-size: 13px;
		display: flex;
		flex-direction: column;
		justify-content: center;
		align-items: center;
		transition: border 0.4s linear;

		p {
			margin: 0;
		}

		&--active {
			border: 1px dashed $blue;
		}

		&::before {
			content: '';
			z-index: -5;
			position: absolute;
			top: -25px;
			left: -25px;
			padding: 24px;
			border-radius: 3px;
			display: block;
			border: 1px solid $gray;
			width: 100%;
			height: 100%;
			transition: border-color 0.3s linear;
		}

		&--active::before {
			border-color: $blue;
		}

		&::after {
			content: '';
			position: absolute;
			top: -20px;
			left: -20px;
			padding: 5px;
			border-radius: 3px;
			display: block;
			border: 15px solid $white;
			width: 100%;
			height: 100%;
			transition: border-color 0.3s linear;
		}

		&--active::after {
			border-color: $primary_light_blue;
		}
	}

	.input-wrapper__upload-instruction {
		padding: 5px;
		font-style: italic;
		color: $dark-gray;
	}

	.input-wrapper__upload-status {
		min-height: 32px;
		text-align: center;
	}

	.input-wrapper__field[type='file'] {
		position: relative;
		z-index: 15;
		height: 150px;
		padding: 50px;
		opacity: 0;
	}
}
