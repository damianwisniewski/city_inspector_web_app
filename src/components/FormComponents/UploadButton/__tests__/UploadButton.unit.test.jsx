import React from 'react'
import { configure, shallow, ShallowWrapper } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import toJson from 'enzyme-to-json'

import UploadButton from '../UploadButton'

configure({ adapter: new Adapter() })

describe('<UploadButton />', () => {
	/**
	 * @type {ShallowWrapper}
	 */
	let SelectElement
	const id = 'mockId'
	const label = 'fakeLabel'
	const onAddImages = () => {}
	const acceptsFile = 'image/'

	const initialState = {
		onDropArea: false,
		loading: false,
		loadingFromDropEvent: false,
		loaded: false,
		files: [],
		error: '',
	}

	beforeEach(() => {
		SelectElement = shallow(
			<UploadButton onAddImages={onAddImages} id={id} label={label} acceptsFile={acceptsFile} />,
		)
	})

	test('should render correctly UploadButton', () => {
		expect(toJson(SelectElement)).toMatchSnapshot()
	})

	test('should render correctly UploadButton with loading state equal true', () => {
		SelectElement.setState({
			...initialState,
			loading: true,
		})
		expect(toJson(SelectElement)).toMatchSnapshot()
	})

	test('should render correctly UploadButton with onDropArea state equal true', () => {
		SelectElement.setState({
			...initialState,
			onDropArea: true,
		})
		expect(toJson(SelectElement)).toMatchSnapshot()
	})

	test('should render correctly UploadButton with loaded state equal true', () => {
		SelectElement.setState({
			...initialState,
			loaded: true,
			files: ['file1', 'file2', 'file3'],
		})
		expect(toJson(SelectElement)).toMatchSnapshot()
	})

	test('should render correctly UploadButton with error state', () => {
		SelectElement.setState({
			...initialState,
			error: 'fake error message',
		})
		expect(toJson(SelectElement)).toMatchSnapshot()
	})
})
