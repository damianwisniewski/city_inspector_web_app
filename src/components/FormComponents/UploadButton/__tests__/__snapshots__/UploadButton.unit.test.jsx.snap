// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UploadButton /> should render correctly UploadButton 1`] = `
<div
  className="input-wrapper input-wrapper--upload"
>
  <input
    accept="image/*"
    className="input-wrapper__field"
    id="mockId"
    multiple={true}
    onChange={[Function]}
    onDragEnter={[Function]}
    onDragLeave={[Function]}
    onDragOver={[Function]}
    onDrop={[Function]}
    type="file"
  />
  <label
    className="input-wrapper__upload-label "
    htmlFor="mockId"
  >
    fakeLabel
    <p
      className="input-wrapper__upload-instruction"
    >
      Kliknij w pole lub przeciągnij plik
    </p>
    <div
      className="input-wrapper__upload-status"
    />
  </label>
</div>
`;

exports[`<UploadButton /> should render correctly UploadButton with error state 1`] = `
<div
  className="input-wrapper input-wrapper--upload"
>
  <input
    accept="image/*"
    className="input-wrapper__field"
    id="mockId"
    multiple={true}
    onChange={[Function]}
    onDragEnter={[Function]}
    onDragLeave={[Function]}
    onDragOver={[Function]}
    onDrop={[Function]}
    type="file"
  />
  <label
    className="input-wrapper__upload-label "
    htmlFor="mockId"
  >
    fakeLabel
    <p
      className="input-wrapper__upload-instruction"
    >
      Kliknij w pole lub przeciągnij plik
    </p>
    <div
      className="input-wrapper__upload-status"
    >
      <Status
        message="fake error message"
        position="horizontal"
        type="error"
      />
    </div>
  </label>
</div>
`;

exports[`<UploadButton /> should render correctly UploadButton with loaded state equal true 1`] = `
<div
  className="input-wrapper input-wrapper--upload"
>
  <input
    accept="image/*"
    className="input-wrapper__field"
    id="mockId"
    multiple={true}
    onChange={[Function]}
    onDragEnter={[Function]}
    onDragLeave={[Function]}
    onDragOver={[Function]}
    onDrop={[Function]}
    type="file"
  />
  <label
    className="input-wrapper__upload-label "
    htmlFor="mockId"
  >
    fakeLabel
    <p
      className="input-wrapper__upload-instruction"
    >
      Kliknij w pole lub przeciągnij plik
    </p>
    <div
      className="input-wrapper__upload-status"
    >
      <Status
        message="Załadowano 3 pliki"
        position="horizontal"
        type="correct"
      />
    </div>
  </label>
</div>
`;

exports[`<UploadButton /> should render correctly UploadButton with loading state equal true 1`] = `
<div
  className="input-wrapper input-wrapper--upload"
>
  <input
    accept="image/*"
    className="input-wrapper__field"
    id="mockId"
    multiple={true}
    onChange={[Function]}
    onDragEnter={[Function]}
    onDragLeave={[Function]}
    onDragOver={[Function]}
    onDrop={[Function]}
    type="file"
  />
  <label
    className="input-wrapper__upload-label "
    htmlFor="mockId"
  >
    fakeLabel
    <p
      className="input-wrapper__upload-instruction"
    >
      Kliknij w pole lub przeciągnij plik
    </p>
    <div
      className="input-wrapper__upload-status"
    >
      <Loader
        id="mockId-loader"
        size="small"
      />
    </div>
  </label>
</div>
`;

exports[`<UploadButton /> should render correctly UploadButton with onDropArea state equal true 1`] = `
<div
  className="input-wrapper input-wrapper--upload"
>
  <input
    accept="image/*"
    className="input-wrapper__field"
    id="mockId"
    multiple={true}
    onChange={[Function]}
    onDragEnter={[Function]}
    onDragLeave={[Function]}
    onDragOver={[Function]}
    onDrop={[Function]}
    type="file"
  />
  <label
    className="input-wrapper__upload-label input-wrapper__upload-label--active"
    htmlFor="mockId"
  >
    fakeLabel
    <p
      className="input-wrapper__upload-instruction"
    >
      Kliknij w pole lub przeciągnij plik
    </p>
    <div
      className="input-wrapper__upload-status"
    />
  </label>
</div>
`;
